temporal-testing/src/main/java/io/temporal/internal/testservice/TestWorkflowService.java

```diff
@@ -572,20 +581,46 @@ public void respondWorkflowTaskFailed(
     }
   }
 
+  private Context.CancellableContext deadlineCtx(final Duration dur) {
+    return Context.current()
+        .withDeadline(Deadline.after(dur.toNanos(), TimeUnit.NANOSECONDS), this.bgSched);
+  }
+
+  /**
+   * Execute blocking function but use the supplied Context to trigger an interrupt of the current
+   * thread in the case that the Context is cancelled.
+   *
+   * @param ctx the context that may be canceled while this thread is blocked in work
+   * @param work the blocking work to perform
+   * @param <T> the result of calling the provided Supplier
+   * @return the result of calling the provided Supplier.
+   */
+  private <T> T withContextInterruption(final Context ctx, final Supplier<T> work) {
+    final Thread thread = Thread.currentThread();
+    final Context.CancellationListener interrupter = context -> thread.interrupt();
+    ctx.addListener(interrupter, this.bgSched);
+    final T out = work.get();
+    ctx.removeListener(interrupter);
+    Thread.interrupted(); // clears interrupted flag
```

Comment:
What if this thread was interrupted not by `interrupter`, but for some other reason? 
Or not only by `interrupter`, but for some other reason? (external shutdownNow on the corresponded thread pool)
We would lose the interrupted flag in this case.

Instead of this "interrupted hack", I would better make an explicit `AtomicBoolean cancelled` that will be set inside `interrupter`. 
And replace untimed .get() call with 
1. Timed get calls to `work`.
2. Checking canceling flag
3. Repeat until timeout is exeeded.

Reply:
That's more or less what the original code did (but it only did it for a duration timeout, not for context cancellation). The reason I didn't double-down on this approach is that the timing and timeout management gets messy. It's difficult (impossible?) to pick a poll time that won't just spin the CPUs but also breaks out of the poll fast enough that the next test execution doesn't fail. All of this is really about getting this poll thread out of the way in time for the next test to execute.

Having said all that, interruption is far from perfect. You're right that this call could clear an interrupt state that wasn't actually created by the Interrupter context listener and for that reason I'd be fine with removing this line altogether. I don't _think_ there's a way to guarantee some sort of interrupt-if-not-already-interrupted call or atomically interrupt the thread _and_ flip a bit in the Interrupter such that we could tell that it was indeed the cause of the interrupt (but if anyone can think of anything ...).

Given the drawbacks, I submit that interruption is the better of two flawed approaches but I think I can be convinced the other way

Reply:
Just though of a way to achieve a preemptible queue poll without using `Thread.interrupt()` or spinning. Working through it now ... ~50% confident it will work :crossed_fingers: